Sources/SwiftSyntaxBuilder/ValidatingSyntaxNodes.swift

```diff
@@ -0,0 +1,58 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SwiftSyntax
+import SwiftDiagnostics
+import SwiftParserDiagnostics
+
+extension SyntaxProtocol {
+  /// If `node` has contains no syntax errors, return `node`, otherwise
+  /// throw an error with diagnostics describing the syntax errors.
+  ///
+  /// This allows clients to e.g. write `try DeclSyntax(validating: "struct Foo {}")`
+  /// which constructs a `DeclSyntax` that's guaranteed to be free of syntax
+  /// errors.
+  public init(validating node: Self) throws {
+    if node.hasError {
+      let diagnostics = ParseDiagnosticsGenerator.diagnostics(for: node)
+      assert(!diagnostics.isEmpty)
+      throw SyntaxStringInterpolationError.diagnostics(diagnostics, tree: Syntax(node))
+    }
+    self = node
+  }
+}
+
+extension Trivia {
+  /// If `trivia` has contains no unexpected trivia, return `trivia`, otherwise
+  /// throw an error with diagnostics describing the unexpected trivia.
+  public init(validating trivia: Trivia) throws {
+    self = trivia
+    if pieces.contains(where: { $0.isUnexpected }) {
+      var diagnostics: [Diagnostic] = []
+      let tree = SourceFileSyntax(statements: [], eofToken: .eof(leadingTrivia: self))
```

Comment:
Little bit weird, but I suppose this is what we were doing before.

Reply:
It does the job of presenting a nice diagnostic so 🤷